### PR TITLE
Add allow_failures flag to Dataset.curve_fit

### LINKED ISSUE #6317: Handle scipy fitting errors in `xarray.curve_fit`
### Is your feature request related to a problem?

I'm trying to fit a bunch of exponential decays to a dataset of dimensions `(Time, experimental_index)`, along the `Time` dimension. This works nicely... except in some cases, `scipy.optimize` raises an error such as `RuntimeError: Optimal parameters not found: The maximum number of function evaluations is exceeded.`. That's, of course, reasonable, except in my use case I'm fine with a bunch of `experimental_index` failing to fit, and I'd still like to access the fits that made it through.

### Describe the solution you'd like

I'm proposing adding a flag for `xarray.curvefit`, for example `allow_failures: bool = False`, that would take this part:

https://github.com/pydata/xarray/blob/f9037c41e36254bfe7efa9feaedd3eae5512bd11/xarray/core/dataset.py#L7738-L7752

and, if active, wrap `popt, pcov = curve_fit(func, x, y, **kwargs)` in a `try/except` block that would output `nan` just as the `not len(y)` conditional clause does if `scipy.optimize.curve_fit` fails with a `RuntimeError`.

### Describe alternatives you've considered

At the moment, I sidestep the issue like this:

```python
fits = []
for experimental_index in timetraces.experimental_index:
    trace = traces.sel(experimental_index=experimental_index)
                                                           
    try:
        fit = trace.curvefit(
            "Time",
            fitfunc,
            p0=(...),
            bounds={...},
        )
    except (RuntimeError, TypeError) as e:
        logger.debug(...)
        continue

    fits.append(fit)

fits = xarray.concat(fits, dim="experimental_index", coords="all")
```
but that seems like a pretty ugly hack...	

### Additional context

I would of course be happy to implement this!

### PR DESCRIPTION
**Pull request by StanczakDominik**
- [x] Closes #6317
- [ ] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

For some reason, my grepping doesn't seem to show any existing tests for `curve_fit`...?

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -7940,7 +7947,14 @@ def _wrapper(Y, *coords_, **kwargs):
                     pcov = np.full([n_params, n_params], np.nan)
                     return popt, pcov
             x = np.squeeze(x)
-            popt, pcov = curve_fit(func, x, y, **kwargs)
+            try:
+                popt, pcov = curve_fit(func, x, y, **kwargs)
+            except RuntimeError:
+                if not allow_failures:
+                    raise
```
Comment (dcherian) [inline xarray/core/dataset.py:7954]:
```suggestion
            except RuntimeError as error:
                if not allow_failures:
                    raise error
```

**Comment by dcherian** [comment]
Thanks @StanczakDominik can you add a test for this behaviour? The existing tests are here: https://github.com/pydata/xarray/blob/f2d0eb49b06f216117f048287e0796c1b6fd0a1d/xarray/tests/test_dataarray.py#L4079

**Comment by StanczakDominik** [comment]
Gah, I kept grepping for `curve_fit` with the underscore! 😅 Thanks for the pointer!

**Comment by headtr1ck** [comment]
@StanczakDominik are you still motivated to work on these tests?

**Comment by mgunyho** [comment]
Hi! I would also like to see this implemented, so I rebased this branch and added a test in #7891.

**Pull request closed without merging**